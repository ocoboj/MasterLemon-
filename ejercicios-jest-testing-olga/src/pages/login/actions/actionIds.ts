export const actionIds = {
  UPDATE_LOGIN_ENTITY_FIELD: 'UPDATE_LOGIN_ENTITY_FIELD',
  UPDATE_LOGIN_FORM_ERRORS: 'UPDATE_LOGIN_FORM_ERRORS',
};
