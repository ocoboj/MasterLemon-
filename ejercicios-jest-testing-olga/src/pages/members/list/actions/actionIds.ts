export const actionIds = {
  UPDATE_MEMBERS: 'UPDATE_MEMBERS',
};
