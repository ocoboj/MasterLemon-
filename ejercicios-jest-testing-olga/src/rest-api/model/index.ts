export * from './loginEntity';
export * from './member';
