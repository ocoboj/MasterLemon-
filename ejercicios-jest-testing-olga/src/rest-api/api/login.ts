import { LoginEntity } from '../model';

export const login = (loginEntity: LoginEntity): Promise<any> => (
  isValidLogin(loginEntity) ?
    Promise.resolve() :
    Promise.reject('Not valid login or password')
);

const isValidLogin = (loginEntity: LoginEntity) => (
  loginEntity.login === 'admin' &&
  loginEntity.password === 'test'
);
