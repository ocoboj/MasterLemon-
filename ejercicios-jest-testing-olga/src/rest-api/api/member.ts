import { Member } from '../model/member';

const baseUrl = 'https://api.github.com/orgs/lemoncode/members';

export const fetchMembers = (): Promise<Member[]> => (
  fetch(baseUrl)
    .then(extractPayload)
);

const extractPayload = (response: Response): Promise<any> => (
  response.ok ?
    response.json() :
    responseError(response)
);

const responseError = (response: Response): Promise<any> => (
  response.json()
    .then((error) => (
      Promise.reject(error.message)
    ))
);
