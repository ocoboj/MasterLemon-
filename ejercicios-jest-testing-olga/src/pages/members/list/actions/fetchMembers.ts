import { actionIds } from './actionIds';
import { Member } from '../../../../rest-api/model';
import * as apiMember from '../../../../rest-api/api/member';

export const fetchMembers = () => (dispatch) => (
  apiMember.fetchMembers()
    .then((members) => dispatch(fetchMembersCompleted(members)))
    .catch(console.log)
);

const fetchMembersCompleted = (members: Member[]) => ({
  type: actionIds.UPDATE_MEMBERS,
  payload: members,
});
