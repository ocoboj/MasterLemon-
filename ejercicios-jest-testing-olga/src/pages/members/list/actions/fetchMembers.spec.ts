import configureStore from 'redux-mock-store';
import reduxThunk from 'redux-thunk';
import * as apiMember from '../../../../rest-api/api/member';
import { Member } from '../../../../rest-api/model';
import { actionIds } from './actionIds';
import { fetchMembers } from './fetchMembers';

const middlewares = [reduxThunk];
const getMocKStore = configureStore(middlewares);

describe('members/list/actions/fetchMembers specs', () => {

  it('should dispatch action UPDATE_MEMBERS and payload with members', (done) => {

    const members: Member[] = [
      {
        id: 1,
        login: 'test login',
        avatar_url: 'test url',
      },
    ];

    const fetchMember = jest.spyOn(apiMember, 'fetchMembers')
      .mockResolvedValue(members);

    const store = getMocKStore();
    store.dispatch<any>(fetchMembers())
      .then(() => {
        const expectAction = store.getActions()[0];
        expect(expectAction.type).toEqual(actionIds.UPDATE_MEMBERS);
        expect(expectAction.type).toMatchSnapshot();
        expect(expectAction.payload).toEqual(members);
        expect(expectAction.payload).toMatchSnapshot();
        done();
      });
  });

  it('should call to apiMember', async () => {

    const fetchMember = jest.spyOn(apiMember, 'fetchMembers').mockResolvedValue([]);

    const store = getMocKStore();
    await store.dispatch<any>(fetchMembers());

    expect(fetchMember).toHaveBeenCalled();
    expect(fetchMember).toMatchSnapshot();
  });

});