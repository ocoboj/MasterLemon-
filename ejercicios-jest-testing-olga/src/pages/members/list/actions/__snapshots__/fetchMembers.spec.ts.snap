// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`members/list/actions/fetchMembers specs should call to apiMember 1`] = `
[MockFunction] {
  "calls": Array [
    Array [],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`members/list/actions/fetchMembers specs should dispatch action UPDATE_MEMBERS and payload with members 1`] = `"UPDATE_MEMBERS"`;

exports[`members/list/actions/fetchMembers specs should dispatch action UPDATE_MEMBERS and payload with members 2`] = `
Array [
  Object {
    "avatar_url": "test url",
    "id": 1,
    "login": "test login",
  },
]
`;
